Source/WebCore/html/URLRegistry.cpp

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2022 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "URLRegistry.h"
+
+#include <wtf/Lock.h>
+#include <wtf/NeverDestroyed.h>
+#include <wtf/Vector.h>
+
+namespace WebCore {
+
+static Lock allRegistriesListLock;
+static Vector<URLRegistry*>& allRegistriesList() WTF_REQUIRES_LOCK(allRegistriesListLock)
+{
+    static NeverDestroyed<Vector<URLRegistry*>> list;
+    return list;
+}
+
+Vector<URLRegistry*> URLRegistry::allRegistries()
+{
+    Vector<URLRegistry*> copy;
+    {
+        Locker locker { allRegistriesListLock };
+        copy = allRegistriesList();
+    }
+    return copy;
+}
+
+URLRegistry::URLRegistry()
+{
+    Locker locker { allRegistriesListLock };
+    allRegistriesList().append(this);
+}
+
+URLRegistry::~URLRegistry()
+{
+    RELEASE_ASSERT_NOT_REACHED(); // All our registries are singleton objects.
```

Comment:
Wish we could delete the destructor instead of implementing it. Would be nicer to get a compile or link time error if we destroy by accident.

Reply:
It would be nice but it's not possible here right? We want to use polymorphism so we need the base class to have a destructor that is virtual.

Reply:
Looks like clang at least is letting me delete the destructor. It is usually not good practice to not have a virtual destructor when having virtual functions (and I thought the compiler would warn about that) but it should be OK here given that this is never deleted. Should I do it?

Reply:
You probably should not. I'd think the destructor would be needed for the vtable used when constructing an object of this class, for example, even if it's never called.